File: coding/src/com/algos/li12_trie/revised/L1268_SearchSuggestionsSystem.java
package  com.algos.li12_trie.revised;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import  com.algos.li30_model.TrieNode;

public class L1268_SearchSuggestionsSystem {

	public static void main(String[] args) {
		L1268_SearchSuggestionsSystem obj = new L1268_SearchSuggestionsSystem();
		
		String[] products = new String[] {"mobile","mouse","moneypot","monitor","mousepad"};
		String searchWord = "mouse";
		System.out.println(obj.suggestedProducts(products, searchWord));
	}

	/*
	 * Problem: https://leetcode.com/problems/search-suggestions-system/
	 * Companies: Amazon
     * Idea: Sort words first as we need to return in lexicographic order
     * Whild building the trie maintain the words at each node. Maintain the words in char trie node.
     * 
	 * */
	public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);      

        TrieNode root = new TrieNode();
        for(String product: products){
            TrieNode curr = root;
            for(char ch: product.toCharArray()){
                if(!curr.map.containsKey(ch)){
                    curr.map.put(ch, new TrieNode());
                }
                curr = curr.map.get(ch);
                if(curr.suggestions.size() < 3)
                    curr.suggestions.add(product);
            }            
        }
        
        List<List<String>> result = new ArrayList<>();
        TrieNode curr = root;
        for(char ch: searchWord.toCharArray()){
            if(curr != null && curr.map.containsKey(ch)){
                curr = curr.map.get(ch);
                result.add(curr.suggestions);
            }else{
                curr = null;
                result.add(Arrays.asList());
            }
        }
        
        return result;
    }
}

File: dsa/concepts/12_tries/python/l1268_search_suggestions_system.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.suggestions = []


"""
Problem: https://leetcode.com/problems/search-suggestions-system/
Companies: Amazon
Idea:
Sort words first as we need to return results in lexicographic order.
While building the trie, maintain up to 3 words at each node.
For each prefix of searchWord, fetch suggestions from the trie.
"""


def suggested_products(products, search_word):
    products.sort()

    root = TrieNode()

    # build trie and store suggestions
    for product in products:
        curr = root
        for ch in product:
            if ch not in curr.children:
                curr.children[ch] = TrieNode()
            curr = curr.children[ch]

            if len(curr.suggestions) < 3:
                curr.suggestions.append(product)

    result = []
    curr = root

    # traverse trie for search word
    for ch in search_word:
        if curr and ch in curr.children:
            curr = curr.children[ch]
            result.append(curr.suggestions)
        else:
            curr = None
            result.append([])

    return result


if __name__ == "__main__":
    products = ["mobile", "mouse", "moneypot", "monitor", "mousepad"]
    search_word = "mouse"

    print(suggested_products(products, search_word))
